src/nupic/regions/knn_classifier_region.py

```diff
@@ -1194,6 +1202,99 @@ def getOutputElementCount(self, name):
       raise Exception('Unknown output: ' + name)
 
 
+  @classmethod
+  def read(cls, proto):
+    if proto.version != KNNClassifierRegion.__VERSION__:
+      raise RuntimeError("Invalid KNNClassifierRegion Version")
+
+    instance = object.__new__(cls)
+    instance.version = proto.version
+    instance._knn = knn_classifier.KNNClassifier.read(proto.knn)
+    instance.knnParams = proto.knnParams.to_dict()
+    instance._rgen = Random()
+    instance._rgen.read(proto.rgen)
+    instance.verbosity = proto.verbosity
+    instance._firstComputeCall = proto.firstComputeCall
+    instance.keepAllDistances = proto.keepAllDistances
+    instance.learningMode = proto.learningMode
+    instance.inferenceMode = proto.inferenceMode
+    instance._doSphering = proto.doSphering
+    instance.outputProbabilitiesByDist = proto.outputProbabilitiesByDist
+    instance._epoch = proto.epoch
+    instance.maxStoredPatterns = proto.maxStoredPatterns
+    instance.maxCategoryCount = proto.maxCategoryCount
+    instance._bestPrototypeIndexCount = proto.bestPrototypeIndexCount
+    instance.acceptanceProbability = proto.acceptanceProbability
+    instance._useAuxiliary = proto.useAuxiliary
+    instance._justUseAuxiliary = proto.justUseAuxiliary
+    instance._protoScoreCount = proto.protoScoreCount
+    instance.confusion = numpy.array(proto.confusion, dtype=numpy.int32)
+    instance._normOffset = numpy.array(proto.normOffset, dtype=numpy.float32)
+    instance._normScale = numpy.array(proto.normScale, dtype=numpy.float32)
+    instance._samples = numpy.array(proto.normScale, dtype=numpy.float32)
+    instance._labels = numpy.array(proto.labels, dtype=numpy.int32)
+    instance._partitions = numpy.array(proto.partitions, dtype=numpy.int32)
+    instance._protoScores = numpy.array(proto.partitions, dtype=numpy.float32)
+    instance._categoryDistances = numpy.array(proto.categoryDistances,
+                                             dtype=numpy.float32)
+
+    instance._scanInfo = None
+    instance._accuracy = None
+    instance._tapFileIn = None
+    instance._tapFileOut = None
+
+    return instance
+
+
+  def write(self, proto):
+    proto.version = self.version
+
+    # Convert 'NoneType' to zero. See 'getParameter'
+    knnParams = self.knnParams.copy()
+    v = knnParams["numSVDSamples"]
+    knnParams["numSVDSamples"] = v if v is not None else 0
+    v = knnParams["numSVDDims"]
+    knnParams["numSVDDims"] = v if v is not None else 0
+    v = knnParams["fractionOfMax"]
+    knnParams["fractionOfMax"] = v if v is not None else 0
+    proto.knnParams = KNNClassifierParamsProto.new_message(**knnParams)
```

Comment:
Can you just do `proto.knnParams = knnParams`? [This page](https://jparyani.github.io/pycapnp/quickstart.html#dictionaries) seems to suggest that you can. And I worry a little bit about explicitly creating a new message for a sub-struct.